File: packages/topo-benchmark/src/topo_benchmark/report.py
"""Generate human-readable benchmark reports."""

from __future__ import annotations


def generate_summary(
    scorecard: dict,
    dimension_details: dict[str, dict] | None = None,
    baseline_results: dict[str, dict] | None = None,
    comparison: dict | None = None,
) -> str:
    """Generate a markdown summary report."""
    lines: list[str] = []
    lines.append("# Benchmark Report")
    lines.append("")
    lines.append(f"**Tier:** {scorecard['tier']}")
    lines.append(f"**Split:** {scorecard['split']}")
    lines.append(f"**Overall Primary Score:** {scorecard['overall_primary']:.4f}")
    lines.append(f"**Promotion Decision:** {scorecard['promotion_decision']}")
    lines.append("")

    # Dimensions table
    lines.append("## Dimension Scores")
    lines.append("")
    lines.append("| Dimension | Score |")
    lines.append("|-----------|-------|")
    for dim, score in scorecard.get("dimensions", {}).items():
        lines.append(f"| {dim} | {score:.4f} |")
    lines.append("")

    # Guardrails
    lines.append("## Guardrails")
    lines.append("")
    for name, passed in scorecard.get("guardrails", {}).items():
        status = "PASS" if passed else "FAIL"
        lines.append(f"- {name}: **{status}**")
    lines.append("")

    # Dimension details
    if dimension_details:
        lines.append("## Dimension Details")
        lines.append("")
        for dim, details in dimension_details.items():
            lines.append(f"### {dim}")
            lines.append("")
            for key, value in details.items():
                if key in ("per_case", "guardrails", "baseline_scores"):
                    continue
                if isinstance(value, float):
                    lines.append(f"- {key}: {value:.4f}")
                else:
                    lines.append(f"- {key}: {value}")
            lines.append("")

    # Baseline comparison
    if baseline_results:
        lines.append("## Baseline Comparison")
        lines.append("")
        for name, results in baseline_results.items():
            lines.append(f"### {name}")
            for key, value in results.items():
                if isinstance(value, float):
                    lines.append(f"- {key}: {value:.4f}")
                else:
                    lines.append(f"- {key}: {value}")
            lines.append("")

    # Comparison (if available)
    if comparison:
        lines.append("## Comparison")
        lines.append("")
        lines.append(f"**Overall Delta:** {comparison['overall_delta']:+.4f}")
        lines.append(f"**Promotion:** {comparison['promotion_decision']}")
        lines.append("")
        lines.append("| Dimension | Candidate | Reference | Delta | CI |")
        lines.append("|-----------|-----------|-----------|-------|----|")
        for dim, d in comparison.get("dimensions", {}).items():
            lines.append(
                f"| {dim} | {d['candidate']:.4f} | {d['reference']:.4f} | "
                f"{d['delta']:+.4f} | [{d['ci_low']:+.4f}, {d['ci_high']:+.4f}] |"
            )
        lines.append("")

    return "\n".join(lines)

File: packages/topo-benchmark/src/report.rs
//! Markdown report generation.

use std::collections::HashMap;

use crate::types::*;

/// Generate a markdown summary report.
pub fn generate_summary(
    scorecard: &Scorecard,
    _dimension_details: &HashMap<String, serde_json::Value>,
    _baseline_results: &HashMap<String, serde_json::Value>,
    comparison: Option<&ComparisonResult>,
) -> String {
    let mut out = String::new();

    out.push_str("# Benchmark Report\n\n");
    out.push_str(&format!("Runner version: {}\n\n", scorecard.runner_version));

    // Overall.
    out.push_str("## Overall\n\n");
    out.push_str(&format!(
        "| Metric | Value |\n|---|---|\n| Overall Primary | {:.4} |\n| Decision | {} |\n| Cases | {}/{} passed |\n\n",
        scorecard.overall_primary,
        scorecard.promotion_decision,
        scorecard.cases_passed,
        scorecard.cases_total,
    ));

    // Dimensions.
    out.push_str("## Dimensions\n\n");
    out.push_str("| Dimension | Score |\n|---|---|\n");
    for dim in Dimension::all() {
        if let Some(score) = scorecard.dimensions.get(dim.as_str()) {
            out.push_str(&format!("| {} | {:.4} |\n", dim.as_str(), score));
        }
    }
    out.push('\n');

    // Guardrails.
    out.push_str("## Guardrails\n\n");
    let g = &scorecard.guardrails;
    out.push_str(&format!("| Check | Status |\n|---|---|\n"));
    out.push_str(&format!(
        "| Coverage | {} |\n| Baseline | {} |\n| No Regressions | {} |\n| False Positive | {} |\n| No Anomaly Flood | {} |\n\n",
        pass_fail(g.coverage_ok),
        pass_fail(g.baseline_ok),
        pass_fail(g.no_regressions),
        pass_fail(g.false_positive_ok),
        pass_fail(g.no_anomaly_flood),
    ));

    // Failing cases.
    if !scorecard.failing_cases.is_empty() {
        out.push_str("## Failing Cases\n\n");
        for case in &scorecard.failing_cases {
            out.push_str(&format!("- {case}\n"));
        }
        out.push('\n');
    }

    // Comparison.
    if let Some(comp) = comparison {
        out.push_str("## Comparison\n\n");
        out.push_str(&format!(
            "Overall: {:.4} → {:.4} (Δ {:.4})\n\n",
            comp.reference_overall, comp.candidate_overall, comp.overall_delta
        ));
        out.push_str("| Dimension | Reference | Candidate | Delta | Status |\n|---|---|---|---|---|\n");
        for dim in Dimension::all() {
            if let Some(d) = comp.dimensions.get(dim.as_str()) {
                let status = if d.regressed { "REGRESSION" } else { "OK" };
                out.push_str(&format!(
                    "| {} | {:.4} | {:.4} | {:.4} | {} |\n",
                    dim.as_str(), d.reference, d.candidate, d.delta, status
                ));
            }
        }
        out.push_str(&format!("\nDecision: {}\n", comp.promotion_decision));
    }

    out
}

fn pass_fail(ok: bool) -> &'static str {
    if ok { "PASS" } else { "FAIL" }
}
